### PR TITLE
Implement onContentPrepare

### PR DESCRIPTION
**Pull request by stutteringp0et**
Pull Request for Issue #111  .

### Summary of Changes

Implement onContentPrepare

### Testing Instructions

The easiest test is to create an editor field and within it type an email address.  If the Email Cloaking plugin is enabled, the output should have a cloaked email address.  

### Expected result

it works

### Actual result

it works

### Documentation Changes Required

tell people it works?

### CODE REVIEW

File: admin/helpers/compiler/e_Interpretation.php [hunk 1]
```diff
@@ -1653,13 +1653,18 @@ public function setCustomViewFieldDecode(&$get,$checker,$string,$code,$tab = '')
...
 			{
 				// build decoder string
-				$fieldUikit .= PHP_EOL."\t".$tab."\t//".$this->setLine(__LINE__)." Make sure the content prepare plugins fire on ".$field." (TODO)";
-				$fieldUikit .= PHP_EOL."\t".$tab."\t".$string."->".$field." = JHtml::_('content.prepare',".$string."->".$field.");";
+				if(!$runplugins) {
+					$runplugins  = PHP_EOL."\t".$tab."\tJPluginHelper::importPlugin('content');\n";
+					$runplugins .= PHP_EOL."\t".$tab."\t".'$dispatcher = JEventDispatcher::getInstance();';
+				}
+				$fieldUikit .= PHP_EOL."\t".$tab."\t//".$this->setLine(__LINE__)." Make sure the content prepare plugins fire on ".$field;
+				$fieldUikit .= PHP_EOL."\t".$tab."\t".'$dispatcher->trigger("onContentPrepare",array($this->_context,&'.$string.'->'.$field.',$item->params));';
```
Comment (Llewellynvdm) [inline admin/helpers/compiler/e_Interpretation.php]: Why did you add `$item->params` to every field check?

File: admin/helpers/compiler/e_Interpretation.php [hunk 2]
```diff
@@ -1653,13 +1653,18 @@ public function setCustomViewFieldDecode(&$get,$checker,$string,$code,$tab = '')
...
+		$runplugins = false;
 		foreach ($checker as $field => $array)
 		{
 			if (strpos($get['selection']['select'], $field) !== false)
 			{
 				// build decoder string
-				$fieldUikit .= PHP_EOL."\t".$tab."\t//".$this->setLine(__LINE__)." Make sure the content prepare plugins fire on ".$field." (TODO)";
-				$fieldUikit .= PHP_EOL."\t".$tab."\t".$string."->".$field." = JHtml::_('content.prepare',".$string."->".$field.");";
+				if(!$runplugins) {
+					$runplugins  = PHP_EOL."\t".$tab."\tJPluginHelper::importPlugin('content');\n";
```
Comment (Llewellynvdm) [inline admin/helpers/compiler/e_Interpretation.php]: Why have a new line `\n` please remove, and make sure to only use `PHP_EOL` to insure it works on windows.

Comment (Llewellynvdm) [reply admin/helpers/compiler/e_Interpretation.php]: You realize it will place it in a loop... I mean this code is place inside the loop of items. So it will load the dispatcher ever time around. I think we must move this dispatcher out of the loop.

File: admin/helpers/compiler/e_Interpretation.php (hunk 1 above)
Comment (stutteringp0et) [reply admin/helpers/compiler/e_Interpretation.php]: Each trigger is the plugin running on a different field.  The plugin event doesn't retain params from the previous plugin run - so it's included every time the plugin event is called.

File: admin/helpers/compiler/e_Interpretation.php (hunk 2 above)
Comment (stutteringp0et) [reply admin/helpers/compiler/e_Interpretation.php]: I inserted multiple editors into my test and the dispatcher was created only once.  Even if it's created more than once, it won't make a difference to the dispatcher - it's a singleton, it's the same object.

newline removed

**Comment by Llewellynvdm** [comment]
I have merged it into staging branch, next time please do target the master branch but the staging, thanks!

File: admin/helpers/compiler/e_Interpretation.php (hunk 1 above)
Comment (Llewellynvdm) [reply admin/helpers/compiler/e_Interpretation.php]: Yes I understand that, but you don't understand that the params variables are not aways on the page. So to pass them without a check is bound to give errors. 

This onContnentPrepare is used in the dynamicGets and as you know the user is in full control of what data is retrieved. So if there is not params best just pass a null.

I will fix this, just that you know in all we do in compiler we are never sure about any value, we must always check before we use value.

Comment (Llewellynvdm) [reply admin/helpers/compiler/e_Interpretation.php]: Even the `_context` value is not always there since we are moving between different modules `JModelList` and `JModelAdmin` and `JModelItem` and every one deals with `_context` in its own way. Like in `JModelList` context is found in [$this->context](https://github.com/Llewellynvdm/joomla-cms/blob/staging/libraries/src/MVC/Model/ListModel.php#L37) not `$this->_context` so we must check in what model this code is going before we decide the context implementation.

Comment (stutteringp0et) [reply admin/helpers/compiler/e_Interpretation.php]: Your tables always create a params field (I haven't found a way to turn it off).  but now that I look at the dynamic get, it appears that param isn't one of the fields imported by default.  It should be simple enough to (property_exists($item,'params')?$item->params:$this->params) - returning the component params when the item params don't exist.  The plugin is happy to get a params object and you have no error.

$this->context vs $this->_context - That sounds like an inconsistency to be resolved.  Without a context, the plugin activity will be unpredictable.  Some plugins don't care, but if someone wanted to write a plugin to act upon ONLY the output of this JCB extension - they wouldn't be able to identify JCB content.  The context is kind of important.

Here's an example of a plugin operating only on a specific context - this one alters the alias to a base65 conversion on the weblink id - so another plugin can use that alias to create a shortcode:

    public function onContentAfterSave($context, $article, $isNew)
    {
        if ( ($context == "com_weblinks.weblink") && ($isNew === true) ) {

            $db = JFactory::getDbo();
            $query = $db->getQuery(true);

            $anyBase = new \weblinks\AnyBase();
            $alias = $anyBase->encode($article->id);

            $fields = array(
                'alias = "' . $alias . '"'
            );

            $conditions = array(
                'id = ' . $article->id
            );

            $query->update($db->quoteName('#__weblinks'))->set($fields)->where($conditions);

            $db->setQuery($query);
            $db->execute();
        }
    }

Comment (Llewellynvdm) [reply admin/helpers/compiler/e_Interpretation.php]: I have resolved this issues, will push it up to stage soon. I just mentioned these things to help you look deeper into the complexity of all being done.

Here is a little peak at the fix that is coming
```
public function setCustomViewFieldonContentPrepareChecker(&$get, $checker, $string, $code, $tab = '', $context, $params)
{
	$fieldPrepare = '';
	$runplugins = false;
	foreach ($checker as $field => $array)
	{
		if (strpos($get['selection']['select'], $field) !== false)
		{
			// build decoder string
			if(!$runplugins) {
				$runplugins = PHP_EOL.$tab."\t//".$this->setLine(__LINE__)." Load the JEvent Dispatcher";
				$runplugins .= PHP_EOL.$tab."\tJPluginHelper::importPlugin('content');";
				$runplugins .= PHP_EOL.$tab."\t".'$this->_dispatcher = JEventDispatcher::getInstance();';
			}
			$fieldPrepare .= PHP_EOL."\t".$tab."\t//".$this->setLine(__LINE__)." Make sure the content prepare plugins fire on ".$field;
			$fieldPrepare .= PHP_EOL."\t".$tab."\t\$".$field." = new stdClass();";
			$fieldPrepare .= PHP_EOL."\t".$tab."\t\$".$field.'->text =& '.$string.'->'.$field.'; // value must be in text';
			$fieldPrepare .= PHP_EOL."\t".$tab."\t".'$this->_dispatcher->trigger("onContentPrepare",array('.$context.',&$'.$field.','.$params.'));';
		}
	}
	// load dispatcher
	$this->JEventDispatcher = array('###DISPATCHER###' => ($runplugins?:''));
	// return UIKIT fix
	return $fieldPrepare;
}
```
